### PR TITLE
RavenDB-18104 - Sharding - Subscriptions - review sharded subscription handlers

### PR DESCRIPTION
**Pull request by lastav5**
https://issues.hibernatingrhinos.com/issue/RavenDB-18104

### CODE REVIEW

File: src/Raven.Server/Documents/Sharding/Handlers/Processors/Subscriptions/ShardedSubscriptionsHandlerProcessorForGetConnectionDetails.cs [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+        {
+            if (ShardedSubscriptionConnection.Connections.TryGetValue(subscriptionName, out var connection) == false)
+            {
+                HttpContext.Response.StatusCode = (int)HttpStatusCode.NotFound;
+                return null;
+            }
+
+            return new SubscriptionConnectionsDetails()
+            {
+                Results = new() {
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Sharding/Handlers/Processors/Subscriptions/ShardedSubscriptionsHandlerProcessorForGetConnectionDetails.cs]: formatting

File: src/Raven.Server/Documents/Sharding/Subscriptions/ShardedSubscriptionConnection.cs [hunk 2]
```diff
@@ -102,7 +102,7 @@ public override async Task ParseSubscriptionOptionsAsync()
             await base.ParseSubscriptionOptionsAsync();
 
             // we want to limit the batch of each shard, to not hold too much memory if there are other batches while batch is proceed
-            _options.MaxDocsPerBatch = Math.Min(_options.MaxDocsPerBatch / TcpConnection.DatabaseContext.ShardCount, _options.MaxDocsPerBatch);
+            _options.MaxDocsPerBatch = Math.Min((int)Math.Ceiling((double)_options.MaxDocsPerBatch / TcpConnection.DatabaseContext.ShardCount), _options.MaxDocsPerBatch);
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Sharding/Subscriptions/ShardedSubscriptionConnection.cs]: better to wrap all of this in `Math.Max(sth, 1);`

File: src/Raven.Server/Documents/Handlers/Processors/OngoingTasks/OngoingTasksHandlerProcessorForDeleteSubscriptionTasks.cs [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+{
+    internal class OngoingTasksHandlerProcessorForDeleteSubscriptionTasks : AbstractOngoingTasksHandlerProcessorForDeleteSubscriptionTasks<DatabaseRequestHandler, DocumentsOperationContext>
+    {
+        public OngoingTasksHandlerProcessorForDeleteSubscriptionTasks([NotNull] DatabaseRequestHandler requestHandler) : base(requestHandler)
+        {
+        }
+
+        protected override async ValueTask DeleteOngoingTaskAsync()
+        {
+            using (var processor = new OngoingTasksHandlerProcessorForDeleteOngoingTask(RequestHandler))
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Handlers/Processors/OngoingTasks/OngoingTasksHandlerProcessorForDeleteSubscriptionTasks.cs]: 1. OngoingTasksHandlerProcessorForDeleteSubscriptionTasks processor should inherit from the OngoingTasksHandlerProcessorForDeleteOngoingTask processor.
2. Override ExecuteAsync
3. Do the Subscription Task validation
4. Execute base.ExecuteAsync();

Comment (ppekrol) [reply src/Raven.Server/Documents/Handlers/Processors/OngoingTasks/OngoingTasksHandlerProcessorForDeleteSubscriptionTasks.cs]: Same for the other one.

File: src/Raven.Server/Documents/Sharding/Subscriptions/ShardedSubscriptionConnection.cs (hunk 2 above)
Comment (lastav5) [reply src/Raven.Server/Documents/Sharding/Subscriptions/ShardedSubscriptionConnection.cs]: done

File: src/Raven.Server/Documents/Sharding/Handlers/Processors/Subscriptions/ShardedSubscriptionsHandlerProcessorForGetConnectionDetails.cs (hunk 1 above)
Comment (lastav5) [reply src/Raven.Server/Documents/Sharding/Handlers/Processors/Subscriptions/ShardedSubscriptionsHandlerProcessorForGetConnectionDetails.cs]: done

File: src/Raven.Server/Documents/Handlers/Processors/OngoingTasks/OngoingTasksHandlerProcessorForDeleteSubscriptionTasks.cs (hunk 3 above)
Comment (lastav5) [reply src/Raven.Server/Documents/Handlers/Processors/OngoingTasks/OngoingTasksHandlerProcessorForDeleteSubscriptionTasks.cs]: done

File: src/Raven.Client/Documents/Commands/DropSubscriptionConnectionCommand.cs [hunk 4]
```diff
@@ -24,23 +25,37 @@ internal DropSubscriptionConnectionCommand(string name, string workerId) : this(
...
+                path.Append("?id=").Append(_subscriptionTaskId);
+            }
 
-                if (string.IsNullOrEmpty(_workerId) == false)
-                {
-                    path.Append("&workerId=").Append(_workerId);
-                }
+            if (string.IsNullOrEmpty(_workerId) == false)
+            {
+                path.Append("&workerId=").Append(_workerId);
```
Comment (garayx) [inline src/Raven.Client/Documents/Commands/DropSubscriptionConnectionCommand.cs]:
```suggestion
                path.Append("&workerId=").Append(Uri.EscapeDataString(_workerId));
```

File: src/Raven.Server/Documents/Handlers/Processors/Subscriptions/AbstractSubscriptionsHandlerProcessorForGetResend.cs [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+                        subscriptionState =
+                            RequestHandler.ServerStore.Cluster.Subscriptions.ReadSubscriptionStateByName(context, RequestHandler.DatabaseName, subscriptionName);
+                    }
+                    catch (SubscriptionDoesNotExistException)
+                    {
+                        HttpContext.Response.StatusCode = (int)HttpStatusCode.NotFound;
+                        return;
+                    }
+                    
+                    resendItems = SubscriptionStorage.GetResendItems(context, RequestHandler.DatabaseName, subscriptionState.SubscriptionId).ToList();
```
Comment (garayx) [inline src/Raven.Server/Documents/Handlers/Processors/Subscriptions/AbstractSubscriptionsHandlerProcessorForGetResend.cs]: why `ToList()` ? `WriteSubscriptionBatchesStateAsync` expects `IEnumerable`

File: src/Raven.Server/Documents/Operations/GetSubscriptionBatchesStateOperation.cs [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+            public GetSubscriptionBatchesStateCommand(string subscriptionName)
+            {
+                _subscriptionName = subscriptionName;
+            }
+
+            public override bool IsReadRequest => true;
+
+            public override HttpRequestMessage CreateRequest(JsonOperationContext ctx, ServerNode node, out string url)
+            {
+                url = $"{node.Url}/databases/{node.Database}/debug/subscriptions/resend?name={_subscriptionName}";
```
Comment (garayx) [inline src/Raven.Server/Documents/Operations/GetSubscriptionBatchesStateOperation.cs]: need to escape the subscription name

File: src/Raven.Server/Documents/Handlers/Processors/OngoingTasks/AbstractOngoingTasksHandlerProcessorForDeleteSubscriptionTasks.cs [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+        public override async ValueTask ExecuteAsync()
+        {
+            // Note: Only Subscription task needs User authentication, All other tasks need Admin authentication
+            var typeStr = RequestHandler.GetQueryStringValueAndAssertIfSingleAndNotEmpty("type");
+            if (Enum.TryParse<OngoingTaskType>(typeStr, true, out var type) == false)
+                throw new ArgumentException($"Unknown task type: {type}", nameof(type));
+
+            if (type != OngoingTaskType.Subscription)
+                throw new ArgumentException("Only Subscription type can call this method");
+
```
Comment (garayx) [inline src/Raven.Server/Documents/Handlers/Processors/OngoingTasks/AbstractOngoingTasksHandlerProcessorForDeleteSubscriptionTasks.cs]: can be moved to static method?

File: src/Raven.Server/Documents/Handlers/Processors/OngoingTasks/AbstractOngoingTasksHandlerProcessorForPostSubscriptionTasksState.cs [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+
+        protected override Task<(long Index, object Result)> OnUpdateConfiguration(TransactionOperationContext context, object _, string raftRequestId)
+        {
+            // Note: Only Subscription task needs User authentication, All other tasks need Admin authentication
+            var typeStr = RequestHandler.GetQueryStringValueAndAssertIfSingleAndNotEmpty("type");
+            if (Enum.TryParse<OngoingTaskType>(typeStr, true, out var type) == false)
+                throw new ArgumentException($"Unknown task type: {type}", nameof(type));
+
+            if (type != OngoingTaskType.Subscription)
+                throw new ArgumentException("Only Subscription type can call this method");
```
Comment (garayx) [inline src/Raven.Server/Documents/Handlers/Processors/OngoingTasks/AbstractOngoingTasksHandlerProcessorForPostSubscriptionTasksState.cs]: can be moved to static method?

File: test/SlowTests/Sharding/Subscriptions/ShardedSubscriptionSlowTests.cs [hunk 9]
```diff
@@ -624,6 +625,62 @@ public async Task ConcurrentSubscriptions()
...
+                            {
+                                amre2.Set();
+                                await amre.WaitAsync();
+                            }
+                            
+                            con1Docs.Add(item.Id);
+                        }
+                    });
+
+                    await amre2.WaitAsync();
```
Comment (garayx) [inline test/SlowTests/Sharding/Subscriptions/ShardedSubscriptionSlowTests.cs]:
```suggestion
                    await amre2.WaitAsync(_reasonableWaitTime);
```

File: test/SlowTests/Sharding/Subscriptions/ShardedSubscriptionSlowTests.cs [hunk 10]
```diff
@@ -624,6 +625,62 @@ public async Task ConcurrentSubscriptions()
...
+                    var amre = new AsyncManualResetEvent();
+                    var amre2 = new AsyncManualResetEvent();
+                    var t = subscription.Run(async x =>
+                    {
+                        foreach (var item in x.Items)
+                        {
+                            if (item.Id == "user/6")
+                            {
+                                amre2.Set();
+                                await amre.WaitAsync();
```
Comment (garayx) [inline test/SlowTests/Sharding/Subscriptions/ShardedSubscriptionSlowTests.cs]:
```suggestion
                                await amre.WaitAsync(_reasonableWaitTime );
```

File: src/Raven.Client/Documents/Commands/DropSubscriptionConnectionCommand.cs (hunk 4 above)
Comment (lastav5) [reply src/Raven.Client/Documents/Commands/DropSubscriptionConnectionCommand.cs]: done

File: src/Raven.Server/Documents/Handlers/Processors/Subscriptions/AbstractSubscriptionsHandlerProcessorForGetResend.cs (hunk 5 above)
Comment (lastav5) [reply src/Raven.Server/Documents/Handlers/Processors/Subscriptions/AbstractSubscriptionsHandlerProcessorForGetResend.cs]: Put the write inside the read transaction and removed `ToList()`

File: src/Raven.Server/Documents/Operations/GetSubscriptionBatchesStateOperation.cs (hunk 6 above)
Comment (lastav5) [reply src/Raven.Server/Documents/Operations/GetSubscriptionBatchesStateOperation.cs]: done

File: src/Raven.Server/Documents/Handlers/Processors/OngoingTasks/AbstractOngoingTasksHandlerProcessorForDeleteSubscriptionTasks.cs (hunk 7 above)
Comment (lastav5) [reply src/Raven.Server/Documents/Handlers/Processors/OngoingTasks/AbstractOngoingTasksHandlerProcessorForDeleteSubscriptionTasks.cs]: done

File: src/Raven.Server/Documents/Handlers/Processors/OngoingTasks/AbstractOngoingTasksHandlerProcessorForPostSubscriptionTasksState.cs (hunk 8 above)
Comment (lastav5) [reply src/Raven.Server/Documents/Handlers/Processors/OngoingTasks/AbstractOngoingTasksHandlerProcessorForPostSubscriptionTasksState.cs]: done

File: test/SlowTests/Sharding/Subscriptions/ShardedSubscriptionSlowTests.cs (hunk 9 above)
Comment (lastav5) [reply test/SlowTests/Sharding/Subscriptions/ShardedSubscriptionSlowTests.cs]: done

File: test/SlowTests/Sharding/Subscriptions/ShardedSubscriptionSlowTests.cs (hunk 10 above)
Comment (lastav5) [reply test/SlowTests/Sharding/Subscriptions/ShardedSubscriptionSlowTests.cs]: done

**Pull request merged**